### PR TITLE
Feature attribute validation

### LINKED ISSUE #21: Attribute Validation
**Is your feature request related to a problem? Please describe.**
There should be the possibility to implement a validation to fields with max/min length on database

**Describe the solution you'd like**
Add a field to metadata with some validations. simple validations as max_length and min_length could be pre-implemented and more complex validations could be added as function references in the validation metadata tag.

**Describe alternatives you've considered**
The `__setattr__` function can be overridden and can implement any validations needed. 

**Additional context**
https://python-reference.readthedocs.io/en/latest/docs/dunderattr/setattr.html

### PR DESCRIPTION
**Pull request by Thalisson-Alves**
This PR adds the possibility to validate attribute values when setting them, and do some changes in existing code:
- The \_\_post_init__ method was splitted into 2 other methods
  - The parsing values of \_\_post_init__ is in the \_\_setattr__ method
  - The part that ensures that an Entity can not be instanciated is in \_\_new__ method, so it fails before the object is instanciated

It's important to leave the parse of values in the \_\_setattr__ method so we can create the validation function that expects the proper value type. e.g.
```
def is_valid_datetime(value) -> bool:
  return value < datetime(2021, 11, 10)

# This would be validated properly
entity.my_datetime = '2021/11/12 10:10:55'
```

**Test plan (required)**

Unit tests asserting that:
- fields are parsed when setting attributes
- if it raises the InvalidAttributeException when validation fails
- do not raise if the value passes the validation
- raises AttributeError when setting an attribute that doesn't exists
- we can raise custom exceptions when the validation function raises it

**Code formatting**

Check that you follow PEP8.

**Closing issues**

closes #21 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the **CONTRIBUTING** document.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: nova_api/entity.py [hunk 1]
```diff
@@ -45,93 +47,48 @@ class Person(Entity):
...
-            logger.error("Trying to instantiate Entity!")
-            raise NotImplementedError("Abstract class can't be instantiated")
+        field_ = getattr(self, '__dataclass_fields__').get(key)
+        if not field_:
+            raise AttributeError
+
+        parsed_value = self.__try_parse_field_value(field_, value)
+        validation_function = field_.metadata.get('validation', lambda v: True)
+        function = validation_function(parsed_value)
+        print(key, value, function)
```
Comment (MatTerra) [inline nova_api/entity.py:86]:
```suggestion
```
Pode ser utilizado um log aqui, mas não deixe o print

File: nova_api/entity.py [hunk 2]
```diff
@@ -45,93 +47,48 @@ class Person(Entity):
...
+        if not function:
```
Comment (MatTerra) [inline nova_api/entity.py:87]:
```suggestion
        if not validation_function(parsed_value):
```

Comment (MatTerra) [reply nova_api/entity.py:87]: Acho que assim fica bacana, o nome function não é descritivo e nem é utilizado em outros lugares

File: nova_api/exceptions/__init__.py [hunk 3]
```diff
@@ -80,3 +80,9 @@ class NoRowsAffectedException(NovaAPIException):
     message: str = field(default="No rows were affected by the "
                                  "desired operation", init=False)
     error_code: int = field(default=304, init=False)
+
+
+@dataclass
+class InvalidAttributeException(NovaAPIException):
+    status_code: int = field(default=400)
+    message: str = field(default='Invalid attribute value')
```
Comment (MatTerra) [inline nova_api/exceptions/__init__.py:88]:
```suggestion
    status_code: int = field(init=False, default=400)
    message: str = field(init=False, default='Invalid attribute value')
```
Quando tiramos o status_code e o message do init, o `debug` se torna o primeiro argumento, dessa forma, `raise InvalidAttributeException("This is debug")` define corretamente o debug mais facilmente.

**Review by MatTerra — CHANGES_REQUESTED** [review]
This is an awesome feature. The documentation is also nicely done. Please check the SonarCloud inspection and fix some of the code smells

File: nova_api/exceptions/__init__.py (hunk 3 above)
Comment (Thalisson-Alves) [reply nova_api/exceptions/__init__.py:88]: Bacana, não sabia dessa

**Review by MatTerra — APPROVED** [review]
👍🏽

**Pull request merged**